### ISSUE TITLE
Counter-intuitive behavior when invoking multiple recipes from justfile in subdirectories

### ISSUE THREAD

**Issue by jpbochi** [casey/just#1473]
I'm making use of the [Invoking justfiles in Other Directories](https://just.systems/man/en/chapter_51.html#invoking-justfiles-in-other-directories) feature. My codebase has a justfile on a `ci` subdirectory, and I'm trying to invoke multiple recipes from there in one call to `just`. Here's the observed behavior.

1. Running `just ci/fe-tests` or `just ci/be-tests`, both work as expected.
2. Running `just ci/fe-tests ci/be-tests` results in this error: `Justfile does not contain recipe ``ci/be-tests``.`
3. Running `just ci/fe-tests be-tests` executes both recipes defined on `ci/Justfile`. That "works", but it's very awkward in my opinion.

I know that the manual says that `just` only cares about the first argument having a `/` or not, and even admits this "may seem a little strange." So, in fairness, `just` is doing what is implemented and documented. I suppose this is a [pain point](https://github.com/casey/just/labels/pain%20point), not a [bug](https://github.com/casey/just/labels/bug).

I have two suggestions to improve this situation:

1. Change the error message to include a full relative path (or perhaps an absolute path) to the Just file. To be clear, say `ci/Justfile does not contain recipe …` rather than `Justfile does not contain recipe …`.
2. Expect all recipes to repeat the path prefix (i.e. `ci/` in my case). I believe this is more intuitive than the current implementation. To be clear, this means that:
   1. `just ci/fe-tests ci/be-tests` would run both recipes from `ci/Justfile`.
   2. `just ci/fe-tests be-tests` would result in an error.
   3. `just ci/fe-tests other-dir/do-smth` would also result in an error saying that running recipes from different justfiles is not supported. Of course, it would be neat to support that, but it's okay to explicitly say it isn't.
   

Related issues: https://github.com/casey/just/issues/333, https://github.com/casey/just/issues/665.

**Comment by casey** [comment]
I agree that it's an awkward situation! Some thoughts:

I think having to repeat the prefix would be clumsy. If you want to run `foo bar and baz` in `some/path/to/a/directory`, you'd have to type:

```
just some/path/to/a/directory/foo some/path/to/a/directory/bar some/path/to/a/directory/baz
```

Instead of:

```
just some/path/to/a/directory/foo bar baz
```

Including the full relative path in the error message seems reasonable. Especially as just gets things like fallback to parent directories and includes, where it might be ambiguous which justfile an error message is talking about.

Checking for recipe names after the first that contain a `.` or `/` and returning an error saying that this is only supported for the first recipe would be a nice improvement to the error message.

**Comment by neunenak** [comment]
What do people think of this PR that tries to provide more ergonomic behavior? https://github.com/casey/just/pull/1475

**Comment by jpbochi** [comment]
@casey I understand your concern about repeating possibly long paths, but I still disagree. Let me try to make a stronger case against it.

1. In your example, people using bash (I'm not sure about zsh and others) could type `just some/path/to/a/directory/{foo,bar,baz}`, without repeating the path. If the path is long, and they are typing this on a terminal, they could also just `cd` there first. If that line is in a script, the typing repetition becomes less of a problem, while the script becomes more readable.
2. IMO, the order of recipes passed to `just` should only affect the order of execution, not the resolution of which recipes are actually executed. This feature creates a unnecessary exception.
3. Requiring the full path in each recipe is more future proof. If you ever want to support running recipes from different justfiles, this feature will become a problem.
4. This feature makes `just` more surprising and less readable. When anyone that is not aware of the feature sees `just a/foo b/bar baz` in a script, I really doubt that they would expect `baz` to be inside `./a/justfile` nor `./b/justfile`.

In essence, this feature is sort of a syntax sugar that is not obvious, is detrimental to usability, and blocks possible improvements.

As a personal anecdote, I'm refraining from using that at all. I'm afraid that writing `just ci/fe-tests be-tests` would confuse my coworkers, and I chose to make two separate calls to `just` instead.

Another relevant anecdote is https://github.com/casey/just/issues/665#issuecomment-869210361. People can reasonably expect that a recipe like `foo: a/bar baz` would work just fine, but it doesn't. They would have to rewrite that like this:

```
foo:
  just a/bar
  just baz
```

And be careful to **not write this**, as it would not do what they probably intend:

```
foo:
  just a/bar baz
```

**Comment by casey** [comment]
Just is a tool that's called frequently as part of a developer's workflow. So, similar to a shell language, some amount of syntactic weirdness is okay, if it makes commands easier to type.

I think it's totally reasonable to avoid this features in justfiles, since it's primarily intended to be used at the command line. Much like I use long flags in scripts, but short flags when I type commands at the command line.

Regarding blocking possible improvements, that ship may have already sailed, given that we can't make backwards incompatible changes now that we're at version 1.0, and backwards incompatible changes to the argument parser are quite hard to make.

**Comment by jpbochi** [comment]
Thanks for explaining your rationale. I don't buy it, but I understand it. As you said, has ship has likely sailed.

With @neunenak's #1475, `just` would support both the weird short syntax and a less surprising longer syntax. That seems like a great compromise to me. Personally, I'll simply opt out from the short syntax, and everyone is happy. :)

**Comment by casey** [comment]
Fixed by #1475.
